core/src/main/java/io/grpc/internal/StatsTraceContext.java

```diff
@@ -0,0 +1,189 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.census.CensusContext;
+import com.google.census.CensusContextFactory;
+import com.google.census.MetricMap;
+import com.google.census.MetricName;
+import com.google.census.RpcConstants;
+import com.google.census.TagKey;
+import com.google.census.TagValue;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Supplier;
+
+import io.grpc.Context;
+import io.grpc.Metadata;
+import io.grpc.Status;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+/**
+ * The stats and tracing information for a call.
+ */
+public final class StatsTraceContext {
+  private enum Side {
+    CLIENT, SERVER
+  }
+
+  private final CensusContext censusContext;
+  private final Stopwatch stopwatch;
+  private final Side side;
+  private final Metadata.Key<CensusContext> censusHeader;
+  private long bytesSent;
+  private long bytesReceived;
+  private boolean callEnded;
+
+  private static final Context.Key<StatsTraceContext> CONTEXT_KEY =
+      Context.key("statsTraceContext");
+
+  private StatsTraceContext(Side side, String methodName, CensusContext parentContext,
+      Supplier<Stopwatch> stopwatchSupplier, Metadata.Key<CensusContext> censusHeader) {
+    this.side = side;
+    TagKey methodTagKey =
+        side == Side.CLIENT ? RpcConstants.RPC_CLIENT_METHOD : RpcConstants.RPC_SERVER_METHOD;
+    this.censusContext = parentContext.with(methodTagKey, new TagValue(methodName));
+    this.stopwatch = stopwatchSupplier.get().start();
+    this.censusHeader = censusHeader;
+  }
+
+  /**
+   * Creates a {@code StatsTraceContext} for an outgoing RPC, using the current CensusContext.
+   */
+  public static StatsTraceContext createOnClient(String methodName,
+      CensusContextFactory censusContextFactory, Supplier<Stopwatch> stopwatchSupplier) {
+    return new StatsTraceContext(Side.CLIENT, methodName, censusContextFactory.getCurrent(),
+        stopwatchSupplier, createCensusHeader(censusContextFactory));
+  }
+
+  /**
+   * Creates a {@code StatsTraceContext} for an incoming RPC, using the CensusContext deserialized
+   * from the headers.
+   */
+  public static StatsTraceContext createOnServer(String methodName,
```

Comment:
"createOnServer" is interesting as it tells you where you should invoke this method. however, it is awkward as when you call a method, you are looking more for what it does. I'd say createServerContext/newServerContext etc is less distracting

Also, seems like a lot of args for a public factory method. seems a lot of the non-variable stuff could be a factory object, right?

ex.
```java
class Factory {
...
  StatsTraceContext newServerContext(String methodName, Metadata headers);
...
}
```